platform-sdk/swirlds-common/src/main/java/com/swirlds/common/scratchpad/Scratchpad.java

```diff
@@ -0,0 +1,387 @@
+import java.util.function.Consumer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A utility for "taking notes" that are preserved across restart boundaries.
+ * <p>
+ * A scratchpad instance is thread safe. All read operations and write operations against a scratchpad are atomic. Any
+ * write that has completed is guaranteed to be visible to all subsequent reads, regardless of crashes/restarts.
+ *
+ * @param <K> the enum type that defines the scratchpad fields
+ */
+public class Scratchpad<K extends Enum<K> & ScratchpadType> {
+
+    private static final Logger logger = LogManager.getLogger(Scratchpad.class);
+
+    /**
+     * The directory where scratchpad files are written. Files are written with the format "0.scr", "1.scr", etc., with
+     * the number increasing by one each time a file is written. If multiple files are present, the file with the
+     * highest number is the most recent and is always used. Multiple files will only be present if the platform crashes
+     * after writing a file but before it has the opportunity to delete the previous file.
+     */
+    public static final String SCRATCHPAD_DIRECTORY_NAME = "scratchpad";
+
+    /**
+     * The file extension used for scratchpad files.
+     */
+    public static final String SCRATCHPAD_FILE_EXTENSION = ".scr";
+
+    private final Set<K> fields;
+    private final Map<Integer, K> indexToFieldMap;
+    private final String id;
+
+    private final Map<K, SelfSerializable> data = new HashMap<>();
+    private final AutoClosableLock lock = Locks.createAutoLock();
+    private final Path scratchpadDirectory;
+    private long nextScratchpadIndex;
+
+    private final int fileVersion = 1;
+
+    /**
+     * Create a new scratchpad.
+     *
+     * @param platformContext the platform context
+     * @param selfId          the ID of this node
+     * @param clazz           the enum class that defines the scratchpad fields
+     * @param id              the unique ID of this scratchpad (creating multiple scratchpad instances on the same node
+     *                        with the same unique ID has undefined (and possibly undesirable) behavior. Must not
+     *                        contain any non-alphanumeric characters, with the exception of the following characters:
+     *                        "_", "-", and ".". Must not be empty.
+     */
+    public Scratchpad(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final NodeId selfId,
+            @NonNull final Class<K> clazz,
+            @NonNull final String id) {
+
+        final StateConfig stateConfig = platformContext.getConfiguration().getConfigData(StateConfig.class);
+        scratchpadDirectory = stateConfig
+                .savedStateDirectory()
+                .resolve(SCRATCHPAD_DIRECTORY_NAME)
+                .resolve(Long.toString(selfId.id()))
+                .resolve(id);
+
+        if (id.isEmpty()) {
```

Comment:
need a non null check first

Reply:
The general pattern we have been following is to do a null check if we are passing an object to a method, but treating it as optional if we dereference it in the constructor. Either way we get a nullpointer exception in the constructor that was given invalid arguments, which is the end goal of the null checks.